orbit-2-test/src/main/java/com/babylon/orbit2/Test.kt

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright 2020 Babylon Partners Limited
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.babylon.orbit2
+
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.atLeast
+import com.nhaarman.mockitokotlin2.doAnswer
+import com.nhaarman.mockitokotlin2.spy
+import com.nhaarman.mockitokotlin2.times
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
+import com.nhaarman.mockitokotlin2.whenever
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runBlocking
+import kotlin.test.assertEquals
+import kotlin.test.fail
+
+/*
+Things to verify:
+1. Sequence of states
+2. Sequence of side effects
+3. "loopbacks"
+4. Invocations on dependencies (mocks)
+5. No other interactions
+ */
+
+fun <STATE : Any, SIDE_EFFECT : Any, T : Host<STATE, SIDE_EFFECT>> T.testSpy(
+    initialState: STATE,
+    isolateFlow: Boolean
+): T {
+    val spy = spy(this)
+    val container = TestContainer<STATE, SIDE_EFFECT>(
+        initialState,
+        isolateFlow
+    )
+    doAnswer { container }.whenever(spy).container
```

Comment:
We lose type safety by writing these doAnswer,whenever. Can we switch it to whenever,doAnswer instead? Can use mockito Kotlin infix version too? Same to apply elsewhere if relevant.

Reply:
Could we tidy this code up more anyway? Something like the following

```
return spy(this) {
    on { container } doAnswer {
        TestContainer...
    }
}